File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

input = sys.stdin.readline

def dfs(n):
    dfs_visit[n] = 1
    print(n, end=" ")
    for w in adjl[n]:
        if dfs_visit[w] == 0:
            dfs_visit[w] = 1
            dfs(w)

def bfs(n):
    bfs_visit[n] = 1
    queue = deque([n])

    while queue:
        t = queue.popleft()
        print(t, end=" ")
        for x in adjl[t]:
            if bfs_visit[x] == 0:
                bfs_visit[x] = 1
                queue.append(x)



n, m, v = map(int, input().split())

adjl = [[] for _ in range(n+1)]

dfs_visit = [0]*(n+1)
bfs_visit = [0]*(n+1)

for _ in range(m):
    a, b = map(int, input().split())
    adjl[a].append(b)
    adjl[b].append(a)

for list in adjl:
    list.sort()

dfs(v)
print()
bfs(v)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;


public class Main {
    // main과 dfs 모두 사용할 것들
    public static List<List<Integer>> adjl =  new ArrayList<>();
    // 정점의 개수 n
    public static int n;
    // 간선의 개수 m
    public static int m;
    // 시작 정점 v
    public static int v;

    public static boolean [] visited_d;
    public static boolean [] visited_b;

    public static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        v= Integer.parseInt(st.nextToken());

        visited_d = new boolean[n+1];
        visited_b = new boolean[n+1];

        // 인접 리스트 생성
        for (int i = 0; i < n+1; i++) {
            adjl.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            adjl.get(a).add(b);
            adjl.get(b).add(a);
        }

        for (int i =0; i < n+1; i++) {
            Collections.sort(adjl.get(i));
        }

        dfs(v);
        sb.append("\n");
        bfs(v);
        System.out.println(sb);


    }

    public static void dfs(int root) {
        visited_d[root] = true;
        sb.append(root).append(" ");
        for (int i : adjl.get(root)) {
            if (!visited_d[i]) {
                dfs(i);
            }
        }
    }

    public static void bfs(int root) {
        Queue<Integer> q = new LinkedList<>();
        q.add(root);
        visited_b[root] = true;
        while (!q.isEmpty()) {
            int v = q.poll();
            sb.append(v).append(" ");
            for (int i : adjl.get(v)) {
                if (!visited_b[i]) {
                    visited_b[i] = true;
                    q.add(i);
                }
            }
        }
    }
}